attributes-next.md

```diff
@@ -0,0 +1,249 @@
+- Normal/storage attributes
+
+Deprecated Behavior:
+
+- Array deep merging is removed (must be manually replaced with hash-key merging)
+- Deep merging of normal values with default and override is removed (users who were incorrectly using normal
+  as a third precedence level will need to manually fix)
+- Deep merging of automatic values with default/normal/override is removed (this is likely just namespace
+  collision mistakes and stright-up bugs?)
+- The force_default/force_override levels are now the standard behavior of setting attributes in recipe code, the
+  "force" behavior becomes the standard behavior.
+- Environment files consistently have a higher priority / later merge order than roles.
+
+Merged Precedence Levels (Current API):
+
+| level | lowest  |  | | highest |
+| ---- | ------ | ------ | ---- | ---- |
+| default       | attribute | recipe | environment | role |
+| force_default | attribute | recipe | | |
+| normal | attribute | recipe | | |
+| override | attribute | recipe | role | environment |
+| force_override | attribute | recipe | | |
+| automatic| node |
+
+Merged Precedence Levels (New API):
+
+| level | lowest  |  | | highest |
+| ---- | ------ | ------ | ---- | ---- |
+| default       | attribute | role | environment | recipe
+| override | attribute | role | environment | recipe
+
+## New Syntax for Ohai Attributes
+
+A new sugar for ohai attributes will be introduced:
+
+```ruby
+ohai("platform")
+ohai("cloud_v2", "public_ipv4_addrs", 0)
+```
+
+This will literally be sugar for the `node.automatic.read("platform")` API that has already been
+shipped.
+
+Write access to ohai data will continue through the existing `node.automatic` API.  In a future RFC
+that will be replaced.
+
+## New Syntax for Merged Default and Override Attributes
+
+```ruby
+attr("foo", "bar", "baz")
+```
+
+This is a reader-only that only reads values merged from default and override levels, with hash-only merging.
+
+It is not intended to be identical to the existing `node["foo"]["bar"]["baz"]` although once deprecation warnings
+against the node object are fixed (see below) they should converge to the same semantics.
+
+The order of merging of the hashes will be:
+
+- default (attribute files)
+- role_default
+- env_default
+- force_default (recipe files)
+- override (attribute files)
+- role_override
+- env_override
+- force_override (recipe files)
+
+Note (below) that attributes written in recipe files will go into the force_default/force_override hashes and will not
+mix with the default/override hashes written in attribute files.  Also note that env_default and role_default have been
+swapped for consistency (the ordering will consistently be attributes/roles/environments/recipes).
+
+# New Syntax for Accessing Default and Override Levels
+
+The writer uses `default_set` and `override_set` while the reader uses simply `default` and `override`.
+
+```ruby
+default_set("foo", "bar", "baz").to("quux")
+default_set("foo", "bar") do |n|
```

Comment:
hmmm `default("foo", "bar", "baz") << 'more'` will have to work since we have to return a mash object and it won't be immutable.  we can also return a decorator for `default_set` which implements `#to` in the decorator and then allows for `default_set("foo", "bar", "baz") << 'more'` as well.